packages/core/src/publicCredential/PublicCredential.chain.ts

```diff
@@ -0,0 +1,230 @@
+
+import { hexToU8a, u8aToHex } from '@polkadot/util'
+import { HexString } from '@polkadot/util/types'
+import { ConfigService } from '@kiltprotocol/config'
+import { fromChain as didFromChain } from '@kiltprotocol/did'
+import { validateUri } from '@kiltprotocol/asset-did'
+import { SDKErrors } from '@kiltprotocol/utils'
+
+import { computeId } from './PublicCredential.js'
+
+export interface EncodedPublicCredential {
+  ctypeHash: CTypeHash
+  subject: AssetDidUri
+  claims: HexString
+  authorization: IDelegationNode['id'] | null
+}
+
+/**
+ * Format a [[IPublicCredentialInput]] to be used as a parameter for the blockchain API function.
+
+ * @param publicCredential The public credential to format.
+ * @returns The blockchain-formatted public credential.
+ */
+export function toChain(
+  publicCredential: IPublicCredentialInput
+): EncodedPublicCredential {
+  const { cTypeHash, claims, subject, delegationId } = publicCredential
+
+  const cborSerializedClaims = cborEncode(claims)
+
+  return {
+    ctypeHash: cTypeHash,
+    subject,
+    // FIXME: Using Uint8Array directly fails to encode and decode, I guess because the api object assumes the byte array is SCALE-encoded.
+    claims: u8aToHex(new Uint8Array(cborSerializedClaims)),
+    authorization: delegationId,
+  }
+}
+
+// Flatten any nested batch calls into a single list of calls.
+function flattenCalls(api: ApiPromise, call: Call): Call[] {
+  if (
+    api.tx.utility.batch.is(call) ||
+    api.tx.utility.batchAll.is(call) ||
+    api.tx.utility.forceBatch.is(call)
+  ) {
+    // Inductive case
+    return call.args[0].reduce(
+      (acc: Call[], c: Call) => acc.concat(flattenCalls(api, c)),
+      []
+    )
+  }
+  // Base case
+  return [call]
+}
+
+// Transform a blockchain-formatted public credential into the original [[IPublicCredentialInput]].
+// It throws if what was written on the chain was garbage.
+function credentialInputFromChain(
+  credential: PublicCredentialsCredentialsCredential
+): IPublicCredentialInput {
+  validateUri(credential.subject.toUtf8())
+  return {
+    claims: cborDecode(hexToU8a(credential.claims.toHex())),
+    cTypeHash: credential.ctypeHash.toHex(),
+    delegationId: credential.authorization.unwrapOr(undefined)?.toHex() ?? null,
+    subject: credential.subject.toUtf8() as AssetDidUri,
+  }
+}
+
+// Retrieve a given block and looks into it to find a public credential creation tx that matches the provided credential ID.
+async function retrievePublicCredentialCreationExtrinsicsFromBlock(
+  api: ApiPromise,
+  credentialId: HexString,
+  blockNumber: u64
+): Promise<Extrinsic | null> {
+  const { extrinsics } = await api.derive.chain.getBlockByNumber(blockNumber)
+  return (
+    extrinsics
+      // Filter out failed extrinsics
+      .filter(({ dispatchError }) => dispatchError === undefined)
+      // Filter out extrinsics that don't contain a `CredentialStored` event with the right credential ID
```

Comment:
I believe that self-documenting code is more expressive and more future-proof at the same time. For example, these two lines could be replaced with `const successfulExtrinsics = extrinsics.filter(({ dispatchError }) => dispatchError === undefined)`. Also, would a simple `!dispatchError` work?

Reply:
The same could be applied to the following steps in the chain as well

Reply:
Very nice suggestion! I guess I tried to make it work since understanding how to parse a block has been quite frustrating. Then I simply forgot to make the code clean and efficient. I fixed all of that in https://github.com/KILTprotocol/sdk-js/pull/677/commits/e113cbbc643026f3dadf6e2857bff733dd69d924#diff-3228260bd7a8a1c0fb1e0b94efbc0b692518737b171dda86329b57db740703bbR100.

I do a `.reverse()` and a `.find()` in two different places where I just need to access the last occurrence of an element in the array. Love this suggestion!

Reply:
Nice!